src/providers/ad/ad_subdomains.c

```diff
@@ -852,9 +886,21 @@ static void ad_get_slave_domain_connect_done(struct tevent_req *subreq)
         return;
     }
 
+    root_dom = ads_get_root_domain(state->be_ctx, state->root_attrs);
+    if (root_dom == NULL) {
+        tevent_req_error(req, ERR_DOMAIN_NOT_FOUND);
+        return;
+    }
+
+    root_sdom = sdap_domain_get(state->opts, root_dom);
```

Comment:
I would suggest to move the sdap_domain_get() call into ads_get_root_domain() so that it becomes ads_get_root_sdap_domain().

Reply:
I like the suggestion, but there is another use of the get-root-domain function:
```
1178     state->root_domain_attrs = reply[0];                                                                         
1179     root_domain = ads_get_root_domain(state->be_ctx, reply[0]);
1180     if (root_domain == NULL) {                                                                                   
1181         DEBUG(SSSDBG_OP_FAILURE, "Could not find the root domain\n");                                            
1182         ret = EFAULT;                                                                                            
1183         goto done;                                                                                               
1184     }                                                                                                            
1185     
1186     state->root_id_ctx = ads_get_root_id_ctx(state->be_ctx,                                                      
1187                                              state->sd_ctx->ad_id_ctx,                                           
1188                                              root_domain, state->opts);                                          
1189     if (state->root_id_ctx == NULL) {                                                                            
1190         DEBUG(SSSDBG_OP_FAILURE, "Cannot create id ctx for the root domain\n");                                  
1191         ret = EFAULT;
1192         goto done;                  
1193     }     
```

On line 1188, the root_domain is used directly. So if you agree, I would create the function, but based on the get-domain one and leave both.